internal/pkg/cli/deploy/patch/env.go

```diff
@@ -0,0 +1,126 @@
+}
+
+// EnsureManagerRoleIsAllowedToUpload checks if the environment manager role has the necessary permission to upload 
+// objects and patch the permissions if not.
+func (d *EnvironmentPatcher) EnsureManagerRoleIsAllowedToUpload(bucketName string) error {
+	body, err := d.TemplatePatcher.EnvironmentTemplate(d.Env.App, d.Env.Name)
+	if err != nil {
+		return fmt.Errorf("get environment template for %q: %w", d.Env.Name, err)
+	}
+	ok, err := isManagerRoleAllowedToUpload(body)
+	if err != nil {
+		return err
+	}
+	if ok {
+		return nil
+	}
+	return d.grantManagerRolePermissionToUpload(d.Env.App, d.Env.Name, d.Env.ExecutionRoleARN, body, s3.FormatARN(endpoints.AwsPartitionID, bucketName))
+}
+
+func (d *EnvironmentPatcher) grantManagerRolePermissionToUpload(app, env, execRole, body, bucketARN string) error {
+	// Detect which line number the EnvironmentManagerRole's PolicyDocument Statement is at.
+	// We will add additional permissions after that line.
+	type Template struct {
+		Resources struct {
+			ManagerRole struct {
+				Properties struct {
+					Policies []struct {
+						Document struct {
+							Statements yaml.Node `yaml:"Statement"`
+						} `yaml:"PolicyDocument"`
+					} `yaml:"Policies"`
+				} `yaml:"Properties"`
+			} `yaml:"EnvironmentManagerRole"`
+		} `yaml:"Resources"`
+	}
+
+	var tpl Template
+	if err := yaml.Unmarshal([]byte(body), &tpl); err != nil {
+		return fmt.Errorf("unmarshal environment template to find EnvironmentManagerRole policy statement: %v", err)
+	}
+	if len(tpl.Resources.ManagerRole.Properties.Policies) == 0 {
+		return errors.New("unable to find policies for the EnvironmentManagerRole")
+	}
+	// lines and columns are 1-indexed, so we have to substract one from each.
+	statementLineIndex := tpl.Resources.ManagerRole.Properties.Policies[0].Document.Statements.Line - 1
+	numSpaces := tpl.Resources.ManagerRole.Properties.Policies[0].Document.Statements.Column - 1
+	pad := strings.Repeat(" ", numSpaces)
+
+	// Create the additional permissions needed with the appropriate indentation.
+	permissions := fmt.Sprintf(`- Sid: PatchPutObjectsToArtifactBucket
+  Effect: Allow
+  Action:
+    - s3:PutObject
+    - s3:PutObjectAcl
+  Resource:
+    - %s
+    - %s/*`, bucketARN, bucketARN)
+	permissions = pad + strings.Replace(permissions, "\n", "\n"+pad, -1)
+
+	// Add the new permissions to the body.
+	lines := strings.Split(body, "\n")
+	linesBefore := lines[:statementLineIndex]
+	linesAfter := lines[statementLineIndex:]
+	updatedLines := append(linesBefore, append(strings.Split(permissions, "\n"), linesAfter...)...)
+	updatedBody := strings.Join(updatedLines, "\n")
+
+	// Update the Environment template with the new content.
+	// CloudFormation is the only entity that's allowed to update the EnvManagerRole so we have to go through this route.
+	// See #3556.
+	var errEmptyChangeSet *cloudformation.ErrChangeSetEmpty
+	// o.prog.Start("Update the environment's manager role with permission to upload artifacts to S3")
```

Comment:
wanna just delete these lines?
```suggestion
```

Reply:
Forgot to add the client for `prog`! I uncommented these lines.